File: frontend/src/services/pdc.service.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
/**
 * PDC (Post-Dated Cheque) Management API Service
 * Story 6.3: Post-Dated Cheque (PDC) Management
 *
 * All PDC-related API calls with comprehensive JSDoc documentation
 */

import { apiClient } from '@/lib/api';
import type {
  PDC,
  PDCDetail,
  PDCListItem,
  PDCFilter,
  PDCCreateRequest,
  PDCBulkCreateRequest,
  PDCDepositRequest,
  PDCClearRequest,
  PDCBounceRequest,
  PDCReplaceRequest,
  PDCWithdrawRequest,
  PDCDashboard,
  PDCWithdrawalFilter,
  PDCWithdrawalHistoryItem,
  CreatePDCResponse,
  BulkCreatePDCResponse,
  GetPDCResponse,
  PDCListResponse,
  PDCDashboardResponse,
  PDCStatusActionResponse,
  PDCReplaceResponse,
  TenantPDCHistoryResponse,
  PDCWithdrawalHistoryResponse,
  BankAccountOption,
  BankAccountsResponse,
  PDCHolder,
  PDCHolderResponse
} from '@/types/pdc';

const PDCS_BASE_PATH = '/v1/pdcs';
const TENANTS_BASE_PATH = '/v1/tenants';
const BANK_ACCOUNTS_BASE_PATH = '/v1/bank-accounts';
const COMPANY_PROFILE_BASE_PATH = '/v1/company-profile';

// ============================================================================
// CREATE PDC
// ============================================================================

/**
 * Create a new single PDC
 *
 * @param data - PDC creation data (tenant, cheque details, amount, date)
 *
 * @returns Promise that resolves to the created PDC
 *
 * @throws {ValidationException} When validation fails (400)
 * @throws {EntityNotFoundException} When tenant/lease/invoice not found (404)
 * @throws {DuplicateException} When cheque number already exists for tenant (409)
 * @throws {UnauthorizedException} When JWT token is missing or invalid (401)
 * @throws {ForbiddenException} When user lacks PROPERTY_MANAGER or FINANCE_MANAGER role (403)
 *
 * @example
 * ```typescript
 * const pdc = await createPDC({
 *   tenantId: '550e8400-e29b-41d4-a716-446655440000',
 *   chequeNumber: '123456',
 *   bankName: 'Emirates NBD',
 *   amount: 50000.00,
 *   chequeDate: '2024-02-15'
 * });
 *
 * console.log(pdc.id);
 * console.log(pdc.status); // "RECEIVED"
 * ```
 */
export async function createPDC(data: PDCCreateRequest): Promise<PDC> {
  const response = await apiClient.post<CreatePDCResponse>(PDCS_BASE_PATH, data);
  return response.data.data;
}

/**
 * Create multiple PDCs in a single bulk operation (atomic)
 *
 * @param data - Bulk PDC creation data with array of cheques (1-24 cheques)
 *
 * @returns Promise that resolves to array of created PDCs
 *
 * @throws {ValidationException} When validation fails (400)
 * @throws {EntityNotFoundException} When tenant/lease not found (404)
 * @throws {DuplicateException} When any cheque number already exists for tenant (409)
 * @throws {ForbiddenException} When user lacks required role (403)
 *
 * @example
 * ```typescript
 * const pdcs = await createBulkPDCs({
 *   tenantId: '550e8400-e29b-41d4-a716-446655440000',
 *   leaseId: '660e8400-e29b-41d4-a716-446655440000',
 *   cheques: [
 *     { chequeNumber: '123456', bankName: 'Emirates NBD', amount: 50000, chequeDate: '2024-02-15' },
 *     { chequeNumber: '123457', bankName: 'Emirates NBD', amount: 50000, chequeDate: '2024-03-15' },
 *     { chequeNumber: '123458', bankName: 'Emirates NBD', amount: 50000, chequeDate: '2024-04-15' }
 *   ]
 * });
 *
 * console.log(`Created ${pdcs.length} PDCs`);
 * ```
 */
export async function createBulkPDCs(data: PDCBulkCreateRequest): Promise<PDC[]> {
  const response = await apiClient.post<BulkCreatePDCResponse>(`${PDCS_BASE_PATH}/bulk`, data);
  return response.data.data;
}

// ============================================================================
// LIST PDCS
// ============================================================================

/**
 * Get paginated list of PDCs with filters
 *
 * @param filters - Optional filters (status, tenant, bank, date range)
 *
 * @returns Promise that resolves to paginated list of PDCListItem
 *
 * @throws {UnauthorizedException} When JWT token is missing or invalid (401)
 * @throws {ForbiddenException} When user lacks required role (403)
 *
 * @example
 * ```typescript
 * // Get all DUE PDCs
 * const response = await getPDCs({
 *   status: 'DUE',
 *   page: 0,
 *   size: 20
 * });
 *
 * // Filter by tenant and date range
 * const filtered = await getPDCs({
 *   tenantId: 'tenant-uuid',
 *   fromDate: '2024-01-01',
 *   toDate: '2024-12-31',
 *   sortBy: 'chequeDate',
 *   sortDirection: 'ASC'
 * });
 * ```
 */
export async function getPDCs(filters?: PDCFilter): Promise<PDCListResponse> {
  const params: Record<string, any> = {
    page: filters?.page ?? 0,
    size: filters?.size ?? 20,
    sortBy: filters?.sortBy ?? 'chequeDate',
    sortDirection: filters?.sortDirection ?? 'ASC'
  };

  // Add filters if provided
  if (filters?.search) {
    params.search = filters.search;
  }
  if (filters?.status && filters.status !== 'ALL') {
    if (Array.isArray(filters.status)) {
      params.status = filters.status.join(',');
    } else {
      params.status = filters.status;
    }
  }
  if (filters?.tenantId) {
    params.tenantId = filters.tenantId;
  }
  if (filters?.bankName) {
    params.bankName = filters.bankName;
  }
  if (filters?.leaseId) {
    params.leaseId = filters.leaseId;
  }
  if (filters?.invoiceId) {
    params.invoiceId = filters.invoiceId;
  }
  if (filters?.fromDate) {
    params.fromDate = filters.fromDate;
  }
  if (filters?.toDate) {
    params.toDate = filters.toDate;
  }

  const response = await apiClient.get<PDCListResponse>(PDCS_BASE_PATH, { params });
  return response.data;
}

// ============================================================================
// GET PDC BY ID
// ============================================================================

/**
 * Get PDC details by ID
 *
 * @param id - PDC UUID
 *
 * @returns Promise that resolves to full PDCDetail with relationships
 *
 * @throws {EntityNotFoundException} When PDC not found (404)
 * @throws {UnauthorizedException} When JWT token is missing or invalid (401)
 * @throws {ForbiddenException} When user lacks required role (403)
 *
 * @example
 * ```typescript
 * const pdc = await getPDCById('550e8400-e29b-41d4-a716-446655440000');
 * console.log(pdc.chequeNumber);
 * console.log(pdc.status);
 * console.log(pdc.statusHistory);
 * ```
 */
export async function getPDCById(id: string): Promise<PDCDetail> {
  const response = await apiClient.get<GetPDCResponse>(`${PDCS_BASE_PATH}/${id}`);
  return response.data.data;
}

// ============================================================================
// PDC STATUS ACTIONS
// ============================================================================

/**
 * Mark PDC as deposited
 * Only DUE status PDCs can be deposited
 *
 * @param id - PDC UUID
 * @param data - Deposit details (date, bank account)
 *
 * @returns Promise that resolves to updated PDC
 *
 * @throws {InvalidStatusException} When PDC is not in DUE status (400)
 * @throws {EntityNotFoundException} When PDC or bank account not found (404)
 * @throws {ForbiddenException} When user lacks required role (403)
 *
 * @example
 * ```typescript
 * const updated = await depositPDC(pdcId, {
 *   depositDate: '2024-02-15',
 *   bankAccountId: 'bank-account-uuid'
 * });
 *
 * console.log(updated.status); // "DEPOSITED"
 * ```
 */
export async function depositPDC(id: string, data: PDCDepositRequest): Promise<PDC> {
  const response = await apiClient.patch<PDCStatusActionResponse>(
    `${PDCS_BASE_PATH}/${id}/deposit`,
    data
  );
  return response.data.data;
}

/**
 * Mark PDC as cleared (payment confirmed by bank)
 * Only DEPOSITED status PDCs can be cleared
 * If linked to invoice, automatically records payment
 *
 * @param id - PDC UUID
 * @param data - Clear details (cleared date)
 *
 * @returns Promise that resolves to updated PDC
 *
 * @throws {InvalidStatusException} When PDC is not in DEPOSITED status (400)
 * @throws {EntityNotFoundException} When PDC not found (404)
 * @throws {ForbiddenException} When user lacks required role (403)
 *
 * @example
 * ```typescript
 * const updated = await clearPDC(pdcId, {
 *   clearedDate: '2024-02-18'
 * });
 *
 * console.log(updated.status); // "CLEARED"
 * // If linked to invoice, payment is automatically recorded
 * ```
 */
export async function clearPDC(id: string, data: PDCClearRequest): Promise<PDC> {
  const response = await apiClient.patch<PDCStatusActionResponse>(
    `${PDCS_BASE_PATH}/${id}/clear`,
    data
  );
  return response.data.data;
}

/**
 * Mark PDC as bounced (payment failed)
 * Only DEPOSITED status PDCs can be bounced
 * Triggers notifications to property manager and tenant
 *
 * @param id - PDC UUID
 * @param data - Bounce details (date, reason)
 *
 * @returns Promise that resolves to updated PDC
 *
 * @throws {InvalidStatusException} When PDC is not in DEPOSITED status (400)
 * @throws {EntityNotFoundException} When PDC not found (404)
 * @throws {ForbiddenException} When user lacks required role (403)
 *
 * @example
 * ```typescript
 * const updated = await bouncePDC(pdcId, {
 *   bouncedDate: '2024-02-18',
 *   bounceReason: 'Insufficient Funds'
 * });
 *
 * console.log(updated.status); // "BOUNCED"
 * // Notifications sent to tenant and property manager
 * ```
 */
export async function bouncePDC(id: string, data: PDCBounceRequest): Promise<PDC> {
  const response = await apiClient.patch<PDCStatusActionResponse>(
    `${PDCS_BASE_PATH}/${id}/bounce`,
    data
  );
  return response.data.data;
}

/**
 * Replace a bounced PDC with a new cheque
 * Only BOUNCED status PDCs can be replaced
 * Creates a new PDC linked to the original
 *
 * @param id - Original PDC UUID (bounced)
 * @param data - New cheque details
 *
 * @returns Promise that resolves to both original and replacement PDCs
 *
 * @throws {InvalidStatusException} When PDC is not in BOUNCED status (400)
 * @throws {EntityNotFoundException} When PDC not found (404)
 * @throws {DuplicateException} When new cheque number already exists (409)
 * @throws {ForbiddenException} When user lacks required role (403)
 *
 * @example
 * ```typescript
 * const result = await replacePDC(pdcId, {
 *   newChequeNumber: '654321',
 *   bankName: 'Emirates NBD',
 *   amount: 50000.00,
 *   chequeDate: '2024-03-01'
 * });
 *
 * console.log(result.original.status); // "REPLACED"
 * console.log(result.replacement.status); // "RECEIVED"
 * console.log(result.replacement.originalChequeId); // Points to original
 * ```
 */
export async function replacePDC(id: string, data: PDCReplaceRequest): Promise<PDCReplaceResponse['data']> {
  const response = await apiClient.post<PDCReplaceResponse>(
    `${PDCS_BASE_PATH}/${id}/replace`,
    data
  );
  return response.data.data;
}

/**
 * Withdraw a PDC (return to tenant)
 * Only RECEIVED or DUE status PDCs can be withdrawn
 *
 * @param id - PDC UUID
 * @param data - Withdrawal details (date, reason, optional new payment method)
 *
 * @returns Promise that resolves to updated PDC
 *
 * @throws {InvalidStatusException} When PDC is not in RECEIVED or DUE status (400)
 * @throws {EntityNotFoundException} When PDC not found (404)
 * @throws {ForbiddenException} When user lacks required role (403)
 *
 * @example
 * ```typescript
 * const updated = await withdrawPDC(pdcId, {
 *   withdrawalDate: '2024-02-10',
 *   withdrawalReason: 'Tenant Request',
 *   newPaymentMethod: 'BANK_TRANSFER',
 *   transactionDetails: {
 *     amount: 50000,
 *     transactionId: 'TXN-12345',
 *     bankAccountId: 'bank-account-uuid'
 *   }
 * });
 *
 * console.log(updated.status); // "WITHDRAWN"
 * ```
 */
export async function withdrawPDC(id: string, data: PDCWithdrawRequest): Promise<PDC> {
  const response = await apiClient.patch<PDCStatusActionResponse>(
    `${PDCS_BASE_PATH}/${id}/withdraw`,
    data
  );
  return response.data.data;
}

/**
 * Cancel a PDC (void)
 * Only RECEIVED status PDCs can be cancelled
 *
 * @param id - PDC UUID
 *
 * @returns Promise that resolves to updated PDC
 *
 * @throws {InvalidStatusException} When PDC is not in RECEIVED status (400)
 * @throws {EntityNotFoundException} When PDC not found (404)
 * @throws {ForbiddenException} When user lacks required role (403)
 *
 * @example
 * ```typescript
 * const updated = await cancelPDC(pdcId);
 * console.log(updated.status); // "CANCELLED"
 * ```
 */
export async function cancelPDC(id: string): Promise<PDC> {
  const response = await apiClient.patch<PDCStatusActionResponse>(
    `${PDCS_BASE_PATH}/${id}/cancel`,
    {}
  );
  return response.data.data;
}

// ============================================================================
// DASHBOARD
// ============================================================================

/**
 * Get PDC dashboard data (KPIs, upcoming PDCs, recently deposited)
 *
 * @returns Promise that resolves to PDCDashboard
 *
 * @example
 * ```typescript
 * const dashboard = await getPDCDashboard();
 *
 * // KPIs
 * console.log(`Due this week: ${dashboard.pdcsDueThisWeek.count} (${dashboard.pdcsDueThisWeek.totalValue} AED)`);
 * console.log(`Deposited: ${dashboard.pdcsDeposited.count}`);
 * console.log(`Outstanding: ${dashboard.totalOutstandingValue} AED`);
 * console.log(`Recently bounced: ${dashboard.recentlyBouncedCount}`);
 *
 * // Lists
 * dashboard.upcomingPDCs.forEach(pdc => console.log(pdc.chequeNumber));
 * dashboard.recentlyDepositedPDCs.forEach(pdc => console.log(pdc.chequeNumber));
 * ```
 */
export async function getPDCDashboard(): Promise<PDCDashboard> {
  const response = await apiClient.get<PDCDashboardResponse>(`${PDCS_BASE_PATH}/dashboard`);
  return response.data.data;
}

// ============================================================================
// WITHDRAWAL HISTORY
// ============================================================================

/**
 * Get PDC withdrawal history with filters
 *
 * @param filters - Optional filters (reason, date range)
 *
 * @returns Promise that resolves to paginated withdrawal history
 *
 * @example
 * ```typescript
 * const history = await getWithdrawals({
 *   withdrawalReason: 'Cheque Bounced',
 *   page: 0,
 *   size: 20
 * });
 *
 * history.data.content.forEach(item => {
 *   console.log(`${item.originalChequeNumber}: ${item.withdrawalReason}`);
 * });
 * ```
 */
export async function getWithdrawals(filters?: PDCWithdrawalFilter): Promise<PDCWithdrawalHistoryResponse> {
  const params: Record<string, any> = {
    page: filters?.page ?? 0,
    size: filters?.size ?? 20,
    sortBy: filters?.sortBy ?? 'withdrawalDate',
    sortDirection: filters?.sortDirection ?? 'DESC'
  };

  if (filters?.search) {
    params.search = filters.search;
  }
  if (filters?.withdrawalReason && filters.withdrawalReason !== 'ALL') {
    params.withdrawalReason = filters.withdrawalReason;
  }
  if (filters?.fromDate) {
    params.fromDate = filters.fromDate;
  }
  if (filters?.toDate) {
    params.toDate = filters.toDate;
  }

  const response = await apiClient.get<PDCWithdrawalHistoryResponse>(
    `${PDCS_BASE_PATH}/withdrawals`,
    { params }
  );
  return response.data;
}

/**
 * Export withdrawal history to file
 *
 * @param format - Export format ('pdf' | 'excel')
 * @param filters - Optional filters to apply
 *
 * @returns Promise that resolves to Blob containing file
 *
 * @example
 * ```typescript
 * const blob = await exportWithdrawals('excel', { fromDate: '2024-01-01' });
 * const url = URL.createObjectURL(blob);
 * const a = document.createElement('a');
 * a.href = url;
 * a.download = 'pdc-withdrawals.xlsx';
 * a.click();
 * ```
 */
export async function exportWithdrawals(
  format: 'pdf' | 'excel',
  filters?: PDCWithdrawalFilter
): Promise<Blob> {
  const params: Record<string, any> = { format };

  if (filters?.search) {
    params.search = filters.search;
  }
  if (filters?.withdrawalReason && filters.withdrawalReason !== 'ALL') {
    params.withdrawalReason = filters.withdrawalReason;
  }
  if (filters?.fromDate) {
    params.fromDate = filters.fromDate;
  }
  if (filters?.toDate) {
    params.toDate = filters.toDate;
  }

  const response = await apiClient.get(`${PDCS_BASE_PATH}/withdrawals/export`, {
    params,
    responseType: 'blob'
  });
  return response.data;
}

// ============================================================================
// TENANT PDC HISTORY
// ============================================================================

/**
 * Get PDC history for a specific tenant with bounce rate
 *
 * @param tenantId - Tenant UUID
 * @param page - Page number (0-indexed)
 * @param size - Page size
 *
 * @returns Promise that resolves to tenant PDC history with bounce rate
 *
 * @example
 * ```typescript
 * const history = await getTenantPDCHistory(tenantId);
 * console.log(`Bounce rate: ${history.data.bounceRate}%`);
 * history.data.content.forEach(pdc => {
 *   console.log(`${pdc.chequeNumber}: ${pdc.status}`);
 * });
 * ```
 */
export async function getTenantPDCHistory(
  tenantId: string,
  page: number = 0,
  size: number = 20
): Promise<TenantPDCHistoryResponse> {
  const response = await apiClient.get<TenantPDCHistoryResponse>(
    `${TENANTS_BASE_PATH}/${tenantId}/pdcs`,
    { params: { page, size } }
  );
  return response.data;
}

// ============================================================================
// BANK ACCOUNTS
// ============================================================================

/**
 * Get available bank accounts for PDC deposit
 * Returns list of company bank accounts with masked account numbers
 *
 * @returns Promise that resolves to list of bank account options
 *
 * @example
 * ```typescript
 * const accounts = await getBankAccounts();
 * accounts.forEach(account => {
 *   console.log(account.displayName); // "Emirates NBD - **** **** **** 1234"
 * });
 * ```
 */
export async function getBankAccounts(): Promise<BankAccountOption[]> {
  try {
    const response = await apiClient.get<BankAccountsResponse>(BANK_ACCOUNTS_BASE_PATH);
    return response.data.data;
  } catch {
    // Graceful handling if Story 6.5 not implemented yet
    return [];
  }
}

// ============================================================================
// PDC HOLDER (COMPANY PROFILE)
// ============================================================================

/**
 * Get PDC holder information from company profile
 * Used to display on PDC detail page
 *
 * @returns Promise that resolves to PDC holder info
 *
 * @example
 * ```typescript
 * const holder = await getPDCHolder();
 * console.log(holder.companyName); // "Emirates Property Care FZ-LLC"
 * ```
 */
export async function getPDCHolder(): Promise<PDCHolder> {
  try {
    const response = await apiClient.get<PDCHolderResponse>(COMPANY_PROFILE_BASE_PATH);
    return response.data.data;
  } catch {
    // Graceful fallback if company profile not configured
    return {
      companyName: 'Company Profile Not Configured'
    };
  }
}

// ============================================================================
// PDC SERVICE OBJECT
// ============================================================================

/**
 * PDC service object with all methods
 * Allows both named imports and object-style access
 */
export const pdcService = {
  // Create operations
  createPDC,
  createBulkPDCs,

  // Read operations
  getPDCs,
  getPDCById,
  getPDCDashboard,
  getWithdrawals,
  getTenantPDCHistory,

  // Status actions
  depositPDC,
  clearPDC,
  bouncePDC,
  replacePDC,
  withdrawPDC,
  cancelPDC,

  // Export
  exportWithdrawals,

  // Related data
  getBankAccounts,
  getPDCHolder
};

File: backend/src/main/java/com/ultrabms/service/PDCService.java
package com.ultrabms.service;

import com.ultrabms.dto.pdc.*;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.UUID;

/**
 * Service interface for PDC (Post-Dated Cheque) operations.
 * Handles PDC lifecycle management from receipt to clearance/bounce/withdrawal.
 *
 * Story 6.3: Post-Dated Cheque (PDC) Management
 * AC #25: PDC Service Layer with business logic
 */
public interface PDCService {

    // =================================================================
    // PDC CREATION
    // =================================================================

    /**
     * Create a single PDC
     * AC #2: PDC Registration form with cheque details
     *
     * @param dto       PDCCreateDto with cheque details
     * @param createdBy User UUID who is registering the PDC
     * @return Created PDC response DTO
     */
    PDCResponseDto createPDC(PDCCreateDto dto, UUID createdBy);

    /**
     * Create multiple PDCs in bulk (1-24 cheques per tenant)
     * AC #3: Atomic bulk PDC registration (all-or-nothing)
     * AC #4: Max 24 PDCs per submission
     *
     * @param dto       PDCBulkCreateDto with tenant and cheque entries
     * @param createdBy User UUID who is registering the PDCs
     * @return List of created PDC response DTOs
     */
    List<PDCResponseDto> createBulkPDCs(PDCBulkCreateDto dto, UUID createdBy);

    // =================================================================
    // PDC RETRIEVAL
    // =================================================================

    /**
     * Get PDC by ID with full details
     *
     * @param id PDC UUID
     * @return PDC response DTO
     */
    PDCResponseDto getPDCById(UUID id);

    /**
     * Get paginated list of PDCs with filters
     * AC #6: Search by cheque number, tenant name, status
     * AC #7: Filter by status, date range, bank
     *
     * @param filterDto Filter parameters
     * @param pageable  Pagination parameters
     * @return Page of PDC list DTOs
     */
    Page<PDCListDto> getPDCs(PDCFilterDto filterDto, Pageable pageable);

    /**
     * Get PDCs for a specific tenant
     *
     * @param tenantId Tenant UUID
     * @param pageable Pagination parameters
     * @return Page of PDC list DTOs
     */
    Page<PDCListDto> getPDCsByTenant(UUID tenantId, Pageable pageable);

    /**
     * Get PDCs linked to an invoice
     *
     * @param invoiceId Invoice UUID
     * @return List of PDC list DTOs
     */
    List<PDCListDto> getPDCsByInvoice(UUID invoiceId);

    // =================================================================
    // PDC STATUS TRANSITIONS
    // =================================================================

    /**
     * Deposit a PDC to bank
     * AC #9: Mark PDC as Deposited (DUE → DEPOSITED)
     *
     * @param id          PDC UUID
     * @param dto         PDCDepositDto with deposit details
     * @param depositedBy User UUID performing the deposit
     * @return Updated PDC response DTO
     */
    PDCResponseDto depositPDC(UUID id, PDCDepositDto dto, UUID depositedBy);

    /**
     * Mark PDC as cleared (payment confirmed)
     * AC #10: Mark PDC as Cleared (DEPOSITED → CLEARED)
     * AC #31: Auto-record payment on linked invoice when cleared
     *
     * @param id        PDC UUID
     * @param dto       PDCClearDto with cleared details
     * @param clearedBy User UUID confirming clearance
     * @return Updated PDC response DTO
     */
    PDCResponseDto clearPDC(UUID id, PDCClearDto dto, UUID clearedBy);

    /**
     * Mark PDC as bounced
     * AC #11: Mark PDC as Bounced (DEPOSITED → BOUNCED)
     *
     * @param id        PDC UUID
     * @param dto       PDCBounceDto with bounce details
     * @param bouncedBy User UUID marking as bounced
     * @return Updated PDC response DTO
     */
    PDCResponseDto bouncePDC(UUID id, PDCBounceDto dto, UUID bouncedBy);

    /**
     * Replace a bounced PDC with a new cheque
     * AC #12: Replace Bounced PDC (BOUNCED → REPLACED)
     * Creates new PDC in RECEIVED status linked to original
     *
     * @param id         Original bounced PDC UUID
     * @param dto        PDCReplaceDto with replacement cheque details
     * @param replacedBy User UUID processing replacement
     * @return New replacement PDC response DTO
     */
    PDCResponseDto replacePDC(UUID id, PDCReplaceDto dto, UUID replacedBy);

    /**
     * Withdraw PDC (return to tenant before deposit)
     * AC #13: Withdraw PDC (RECEIVED/DUE → WITHDRAWN)
     * AC #14: Record alternative payment method when withdrawn
     *
     * @param id          PDC UUID
     * @param dto         PDCWithdrawDto with withdrawal details
     * @param withdrawnBy User UUID processing withdrawal
     * @return Updated PDC response DTO
     */
    PDCResponseDto withdrawPDC(UUID id, PDCWithdrawDto dto, UUID withdrawnBy);

    /**
     * Cancel a PDC (void before deposit)
     * AC #15: Cancel PDC (RECEIVED → CANCELLED)
     *
     * @param id          PDC UUID
     * @param cancelledBy User UUID cancelling the PDC
     * @return Updated PDC response DTO
     */
    PDCResponseDto cancelPDC(UUID id, UUID cancelledBy);

    // =================================================================
    // DASHBOARD AND ANALYTICS
    // =================================================================

    /**
     * Get PDC dashboard data with KPIs
     * AC #1: PDC Dashboard with summary KPIs
     *
     * @return PDC dashboard DTO
     */
    PDCDashboardDto getDashboard();

    /**
     * Get tenant's PDC history and statistics
     *
     * @param tenantId Tenant UUID
     * @param pageable Pagination for PDC list
     * @return Tenant PDC history DTO
     */
    TenantPDCHistoryDto getTenantPDCHistory(UUID tenantId, Pageable pageable);

    // =================================================================
    // WITHDRAWAL HISTORY
    // =================================================================

    /**
     * Get withdrawal history page
     * AC #16: Withdrawal history with filters
     *
     * @param pageable Pagination parameters
     * @return Page of withdrawn PDCs
     */
    Page<PDCListDto> getWithdrawalHistory(Pageable pageable);

    // =================================================================
    // UTILITY METHODS
    // =================================================================

    /**
     * Get distinct bank names for filter dropdown
     *
     * @return List of bank names
     */
    List<String> getDistinctBankNames();

    /**
     * Get PDC holder name from company profile
     * AC #5: Display PDC Holder (company legalCompanyName) on form
     *
     * @return Company legal name or default message
     */
    String getPDCHolderName();

    /**
     * Check if cheque number already exists for tenant
     *
     * @param chequeNumber Cheque number
     * @param tenantId     Tenant UUID
     * @return true if cheque number exists
     */
    boolean chequeNumberExists(String chequeNumber, UUID tenantId);

    // =================================================================
    // SCHEDULER SUPPORT
    // =================================================================

    /**
     * Transition RECEIVED PDCs to DUE status when cheque date is within 7 days
     * Called by 6 AM scheduler job
     * AC #8: Automated RECEIVED → DUE transition
     *
     * @return Number of PDCs transitioned
     */
    int transitionReceivedToDue();

    /**
     * Get PDCs due on a specific date for reminder emails
     * Called by 9 AM scheduler job for deposit reminders
     * AC #29: Email reminder for PDCs due for deposit
     *
     * @param reminderDate Date to check
     * @return List of PDCs due on that date
     */
    List<PDCListDto> getPDCsDueForReminder(java.time.LocalDate reminderDate);
}
